content/client-lib-development-guide/features.textile

```diff
@@ -696,3 +701,355 @@ The following default values are configured for the client library:
+  data: Data? // TM2d
+  encoding: String? // TM2e
+  id: String // TM2a
+  name: String? // TM2g
+  timestamp: Time // TM2f
+
+class PresenceMessage
+  action: PresenceAction // TP3b
+  clientId: String // TP3c
+  connectionId: String // TP3d
+  data: Data? // TP3e
+  encoding: String? // TP3f
+  id: String // TP3a
+  timestamp: Time // TP3g
+  memberKey: () -> String // TP3h
+
+class ProtocolMessage:
+  action: ProtocolMessageAction // TR2, TR4a
+  channel: String? // TR4b
+  channelSerial: String? // TR4c
+  connectionDetails: ConnectionDetails? // RSA7b3, RTN19, TR4o
+  connectionId: String? // RTN15c1, TR4d
+  connectionKey: String? // TR4e
+  connectionSerial: Int? // RTN10c, TR4f
+  count: Int? // TR4g
+  error: ErrorInfo? // RTN15c2, TR4h
+  flags: .HAS_PRESENCE & .HAS_BACKLOG ? // RTP1, TR3, TR4i
+  id: String? // TR4b
+  messages: [Message]? // TR4k
+  msgSerial: Int? // RTN7b, TR4j
+  presence: [PresenceMessage]? // TR4l
+  timestamp: Time? // TR4m
+
+enum ProtocolMessageAction:
+  HEARTBEAT // TR2
+  ACK // TR2
+  NACK // TR2
+  CONNECT // TR2
+  CONNECTED // TR2
+  DISCONNECT // TR2
+  DISCONNECTED // TR2
+  CLOSE // TR2
+  CLOSED // TR2
+  ERROR // TR2
+  ATTACH // TR2
+  ATTACHED // TR2
+  DETACH // TR2
+  DETACHED // TR2
+  PRESENCE // TR2
+  MESSAGE // TR2
+  SYNC // TR2
+
+class Connection:
+  embeds EventEmitter<ConnectionState, ConnectionStateChange> // RTN4a, RTN4e
+  errorReason: ErrorInfo? // RTN14a
+  id: String? // RTN8
+  key: String? // RTN9
+  recoveryKey: String? // RTN16b, RTN16c
+  serial: Int // RTN10
+  state: ConnectionState // RTN4d
+  close: () // RTN12
+  connect: () // RTC1b, RTN3, RTN11
+  ping: () -> io // RTN13
+
+enum ConnectionState:
+  INITIALIZED
+  CONNECTING
+  CONNECTED
+  DISCONNECTED
+  SUSPENDED
+  CLOSING
+  CLOSED
+  FAILED
+
+class ConnectionStateChange:
+  current: ConnectionState // TA2
+  previous: ConnectionState // TA2
+  reason: ErrorInfo? // RTN4f, TA3
+  retryIn: Duration? // RTN14d, TA2
+
+class Stats:
+  all: StatsMessageTypes // http://goo.gl/TpIh5I
+  apiRequests: StatsRequestCount // http://goo.gl/lFhZLC
+  channels: StatsResourceCount // http://goo.gl/jfu0q1
+  connections: StatsConnectionTypes // http://goo.gl/72fQ7Z
+  inbound: StatsMessageTraffic // http://goo.gl/8SjQAJ
+  intervalGranularity: StatsIntervalGranularity
+  intervalId: String
+  intervalTime: Time
+  outbound: StatsMessageTraffic // http://goo.gl/8SjQAJ
+  persisted: StatsMessageTypes // http://goo.gl/TpIh5I
+  tokenRequests: StatsRequestCount // http://goo.gl/lFhZLC
+
+enum StatsIntervalGranularity:
+  MINUTE
+  HOUR
+  DAY
+  MONTH
+
+class ErrorInfo:
+  code: Int // TI1
+  message: String // TI1
+  statusCode: Int // TI1
+
+class EventEmitter<Event, Data>:
+  on: ((Data) ->) // RTE4
+  on: (Event, (Data) ->) // RTE4
+  once: ((Data) ->) // RTE4
+  once: (Event, (Data) ->) // RTE4
+  off: () // RTE5
+  off: ((Data) ->) // RTE5
+  off: (Event, (Data) ->) // RTE5
+  emit: (Event, Data?)  // RTE6
```

Comment:
Well in fact I always understood the EventEmitter to be able to emit any number of arguments, but also not emit any.  See https://github.com/ably/ably-ruby/blob/master/lib/ably/modules/event_emitter.rb#L88 and https://github.com/ably/ably-java/blob/master/lib/src/main/java/io/ably/lib/util/EventEmitter.java#L75.  Both allow zero+ arguments. How do you want to approach this so that it's flexible for iOS as well then?